resources: Add source code for riscv-tests

kunpai opened the pull request:
- This code is compatible with m5ops, i.e., ROIs are set using m5_work_begin and m5_work_end.
- Added README.md and CHANGELOG.md for every benchmark to track issues
- Updated base README.md of asmtest directory to show what code has been changed
- riscv-tests benchmarks' source code is now housed inside src/asmtest/benchmarks. Earlier, it used to be in src/riscv-tests

Change-Id: I17572d00b50617bc52e8453ee873b0c2f614f461

kunpai committed:
resources: Add source code for riscv-tests (#1)

- This code is compatible with m5ops, i.e., ROIs are set using m5_work_begin and m5_work_end.
- Added README.md and CHANGELOG.md for every benchmark to track issues
- Updated base README.md of asmtest repository to show what all has changed
- These benchmarks are now housed inside src/asmtest/benchmarks. Earlier, they used to be in src/riscv-tests

powerjg commented on src/asmtest/.gitignore:
Probably should be in a separate commit

powerjg commented on src/asmtest/benchmarks/Makefile:
```diff
@@ -23,32 +23,44 @@ bmarks = \
+RISCV_PREFIX ?= riscv$(XLEN)-linux-gnu-
 RISCV_GCC ?= $(RISCV_PREFIX)gcc
-RISCV_GCC_OPTS ?= -DPREALLOCATE=1 -mcmodel=medany -static -std=gnu99 -O2 -ffast-math -fno-common -fno-builtin-printf -fno-tree-loop-distribute-patterns
+RISCV_GCC_OPTS ?= -DPREALLOCATE=1 -mcmodel=medany -static -std=gnu99 -O2 -ffast-math -fno-common -fno-builtin-printf -fno-tree-loop-distribute-patterns -march=rv$(XLEN)gcv -mabi=lp64d
```
Not sure it  makes sense to allow `XLEN` to change and not update `mabi`

kunpai replied:
This change came from riscv-tests. Should I update mabi to use `lp($XLEN)d` instead?

powerjg commented on src/asmtest/benchmarks/Makefile:
```diff
@@ -23,32 +23,44 @@ bmarks = \
 RISCV_LINK_OPTS ?= -static -nostdlib -nostartfiles -lm -lgcc -T $(src_dir)/common/test.ld
 RISCV_OBJDUMP ?= $(RISCV_PREFIX)objdump --disassemble-all --disassemble-zeroes --section=.text --section=.text.startup --section=.text.init --section=.data
-RISCV_SIM ?= spike --isa=rv$(XLEN)gc
+RISCV_SIM ?= spike
```
Why this change?

kunpai replied:
I was playing around with different compilation options and I forgot to reset it. I will add this back. Thanks for pointing this out!

powerjg commented on src/asmtest/benchmarks/Makefile:
```diff
@@ -23,32 +23,44 @@ bmarks = \
 define compile_template
 $(1).riscv: $(wildcard $(src_dir)/$(1)/*) $(wildcard $(src_dir)/common/*)
-	$$(RISCV_GCC) $$(incs) $$(RISCV_GCC_OPTS) -o $$@ $(wildcard $(src_dir)/$(1)/*.c) $(wildcard $(src_dir)/common/*.c) $(wildcard $(src_dir)/common/*.S) $$(RISCV_LINK_OPTS)
+	$$(RISCV_GCC) -I$(src_dir)/common -static -O2 -o $$@ $(wildcard $(src_dir)/$(1)/*.c) $(wildcard $(src_dir)/$(1)/*.S) $(src_dir)/common/gem5/util/m5/build/riscv/out/libm5.a 
```
This makes an assumption about where gem5/ is. You should make sure to document this or make a variable.

kunpai replied:
I have documented this in the README.md of `src/asmtest`

powerjg commented on src/asmtest/benchmarks/common/crt.S:
```diff
@@ -47,8 +47,8 @@ _start:
-  # enable FPU and accelerator if present
-  li t0, MSTATUS_FS | MSTATUS_XS
+  # enable FPU, vector, and accelerator if present
+  li t0, MSTATUS_FS | MSTATUS_XS | MSTATUS_VS
```
Needed?

kunpai replied:
I am not really sure. This came from `riscv-tests`.

powerjg commented on src/asmtest/benchmarks/common/encoding.h:
Can you explain these changes?

kunpai replied:
Since I did not use the riscv-tests linkers and just compiled with linking to the `src/asmtest/benchmarks/common/` directory and m5ops, they could not find `encoding.h`, which was in a different directory. So, I just put `encoding.h` into the `common` directory. I found trouble using m5ops if I linked with two different directories while compiling.

powerjg reviewed (commented):
A few small comments after going through the first few files.

Overall.. can you explain what you wrote and what came directly from the upstream?

kunpai commented:
> A few small comments after going through the first few files.
> 
> Overall.. can you explain what you wrote and what came directly from the upstream?

Mostly, I changed the compile_template of the Makefile to not use the riscv-tests linker. In the source code, I did not change much, except comment out places where spike magic instructions are being called (and it is not important to the run of the program, Start_Timer(); in dhrystone for example) and replace their setStats calls with m5ops. I have documented what I changed in the README.md of src/asmtest.

powerjg commented:
It would be great if you could separate out your changes from the changes
that were introduced upstream.

The best would be to have 3 commits:
- Moving the files around and deleting unneeded files
- Updating to latest riscv-tests upstream
- Your changes.

If this isn't possible, then we can figure out how to review it as-is.

On Thu, Aug 3, 2023 at 1:58 PM Kunal Pai ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/asmtest/benchmarks/Makefile
> <https://github.com/gem5/gem5-resources/pull/3#discussion_r1283713363>:
>
> >  RISCV_LINK ?= $(RISCV_GCC) -T $(src_dir)/common/test.ld $(incs)
>  RISCV_LINK_OPTS ?= -static -nostdlib -nostartfiles -lm -lgcc -T $(src_dir)/common/test.ld
>  RISCV_OBJDUMP ?= $(RISCV_PREFIX)objdump --disassemble-all --disassemble-zeroes --section=.text --section=.text.startup --section=.text.init --section=.data
> -RISCV_SIM ?= spike --isa=rv$(XLEN)gc
> +RISCV_SIM ?= spike
>
> I was playing around with different compilation options and I forgot to
> reset it. I will add this back. Thanks for pointing this out!
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/gem5/gem5-resources/pull/3#discussion_r1283713363>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAA2YHAIBKRRXQK6NBB52NTXTQGF5ANCNFSM6AAAAAA3DJMCJY>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

kunpai commented:
> It would be great if you could separate out your changes from the changes that were introduced upstream. The best would be to have 3 commits: - Moving the files around and deleting unneeded files - Updating to latest riscv-tests upstream - Your changes. If this isn't possible, then we can figure out how to review it as-is.
> […](#)

Okay, sounds good to me. Let me see what I can do. For now, I will close this pull request.
